### PR TITLE
Add (with-stdin-from ...)

### PR DESCRIPTION
**Pull request by nojb**
This one keeps coming up, and it makes sense since all the infrastructure is already there.

### CODE REVIEW

File: src/action_to_sh.ml [hunk 1]
```diff
@@ -35,8 +35,10 @@ let simplify act =
       loop act (Chdir p :: mkdir p :: acc)
     | Setenv (k, v, act) ->
       loop act (Setenv (k, v) :: acc)
-    | Redirect (outputs, fn, act) ->
+    | Redirect_out (outputs, fn, act) ->
       Redirect (block act, outputs, File fn) :: acc
+    | Redirect_in _ ->
+      Code_error.raise "not yet implemented" []
```
Comment (nojb) [inline src/action_to_sh.ml]: I will fill this one out shortly.

Comment (nojb) [reply src/action_to_sh.ml]: This is done.

File: src/process.ml [hunk 2]
```diff
@@ -92,6 +92,59 @@ module Output = struct
     { t with status = Keep_open }
 end
 
+module Input = struct
```
Comment (rgrinberg) [inline src/process.ml]: I'm not convinced all this duplication is necessary. Would it be possible to repurpose the output module instead? Call it something more generic and make the channel the type variable. We could even use a GADT to make sure that in_channel and out_channel are the only possibilities but that isn't essential.

**Review by rgrinberg — CHANGES_REQUESTED** [review]
Looking good nojb :) Just a few tweaks please.

File: src/action_ast.ml [hunk 3]
```diff
@@ -75,6 +82,11 @@ struct
            and+ t = t
            in
            translate_to_ignore fn Outputs t)
+        ; "with-stdin-from",
+          (let+ fn = path
```
[the remark this answers is not in the document]
Comment (nojb) [reply src/action_ast.ml]: Done, now it requires dune 2.0.

File: test/blackbox-tests/test-cases/with-stdin-from/dune-project [hunk 4]
```diff
@@ -0,0 +1 @@
+(lang dune 1.12)
```
[the remark this answers is not in the document]
Comment (nojb) [reply test/blackbox-tests/test-cases/with-stdin-from/dune-project]: Done.

File: CHANGES.md [hunk 5]
```diff
@@ -59,9 +59,11 @@
 - Add a new config option `sandboxing_preference`, the cli argument `--sandbox`,
   and the dep spec `sandbox` in dune language. These let the user control the level of
   sandboxing done by dune per rule and globally. The rule specification takes precedence.
-  The global configuration merely specifies the default. 
+  The global configuration merely specifies the default.
   (#2213, @aalekseyev, @jdimino)
 
+- Add action (with-stdin-from ...). (#2487, @nojb)
```
[the remark this answers is not in the document]
Comment (nojb) [reply CHANGES.md]: Fixed, thanks.

File: src/process.ml (hunk 2 above)
Comment (nojb) [reply src/process.ml]: The last commit implements this (GADTs included :)).

**Review by rgrinberg — APPROVED** [review]
Awesome work Nicolas, I don't even mind the fix the tests for you this time :)
Let's just make sure to squash away all these meaningless commits.

**Comment by nojb** [comment]
> Awesome work Nicolas, I don't even mind the fix the tests for you this time :)

Thanks! I wrote the code on Windows, where many tests don't run, which is why I missed them.

> Let's just make sure to squash away all these meaningless commits.

Will do.

**Pull request merged**